### PR TITLE
[Diffusion] [Profiling] Add end-to-end profiling support for diffusion serving pipeline

### PR DESCRIPTION
**Pull request by Jasen2201**
# [Diffusion] [Profiling] Add end-to-end profiling support for diffusion serving pipeline

## Summary

This PR adds end-to-end profiling support for the diffusion serving pipeline, enabling developers to capture synchronized `torch.profiler` traces from **both** the HTTP Server (host) process and the GPU Worker process with a single command. The profiling workflow is designed to be consistent with the existing LLM profiling API in SGLang.

## Motivation

Profiling diffusion serving pipelines requires visibility into both the host-side overhead (HTTP request handling, scheduling, data encoding/decoding) and the GPU-side computation (denoising steps, attention, VAE). Previously, there was no unified way to capture traces from both processes simultaneously. This PR provides that capability, making it easy to identify bottlenecks across the full serving stack.

## Usage

**1. Launch the server with profiling environment variables:**

```bash
export SGLANG_TORCH_PROFILER_DIR=./sglang_qwen_profiling
export SGLANG_PROFILE_WITH_STACK=1
export SGLANG_PROFILE_RECORD_SHAPES=1

sglang serve \
    --model-path <model_path> \
    --num-gpus 2 \
    --ulysses-degree 2 \
    --host 0.0.0.0 \
    --port 40000
```

**2. Run the benchmark with profiling:**

```bash
# Warm up (no profiling)
python3 -m sglang.multimodal_gen.benchmarks.bench_serving \
    --backend sglang-image --task image-to-image \
    --dataset vbench --dataset-path <dataset_path> \
    --max-concurrency 1 --num-prompts 1

# Profile run
python3 -m sglang.multimodal_gen.benchmarks.bench_serving \
    --backend sglang-image --task image-to-image \
    --dataset vbench --dataset-path <dataset_path> \
    --max-concurrency 1 --num-prompts 2 \
    --num-inference-steps 8 --guidance-scale 1 \
    --profile
```

**3. Inspect traces in Chrome (`chrome://tracing`) or Perfetto UI:**

The profiler generates two trace files per profiling session:
- `{profile_id}-host.trace.json.gz` — Host process timeline
- `{profile_id}-rank-{rank}.trace.json.gz` — GPU Worker timeline

## Profiling Results

**Host Process Timeline:**

image

**GPU Worker Timeline:**

image

### CODE REVIEW

File: python/sglang/multimodal_gen/benchmarks/bench_serving.py [hunk 1]
```diff
@@ -698,6 +760,16 @@ async def limited_request_func(req, session, pbar):
         else:
             return await request_func(req, session, pbar)
 
+    # Start profiler if --profile is enabled
+    if args.profile:
+        profile_output = await async_request_profile(
+            api_url=f"{args.base_url}/start_profile"
+        )
+        if profile_output.success:
+            print(f"Profiler started: {profile_output.message}")
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/benchmarks/bench_serving.py]: shall we use `logger.info` instead of `print`

File: python/sglang/multimodal_gen/benchmarks/bench_serving.py [hunk 2]
```diff
@@ -309,6 +314,8 @@ def __getitem__(self, idx: int) -> RequestFuncInput:
             height=self.args.height,
             num_frames=self.args.num_frames,
             fps=self.args.fps,
+            num_inference_steps=getattr(self.args, "num_inference_steps", None),
+            guidance_scale=getattr(self.args, "guidance_scale", None),
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/benchmarks/bench_serving.py]: i think we can directly use `self.args.num_inference_steps` because we've already set the default to None

File: python/sglang/multimodal_gen/benchmarks/bench_serving.py [hunk 3]
```diff
@@ -360,8 +367,17 @@ async def async_request_image_sglang(
...
-        # Merge extra parameters
-        for key, value in input.extra_body.items():
+        # Add inference parameters
+        if input.num_inference_steps is not None:
+            data.add_field("num_inference_steps", str(input.num_inference_steps))
+        if input.guidance_scale is not None:
+            data.add_field("guidance_scale", str(input.guidance_scale))
+
+        # Add profiling and other extra parameters
+        extra_params = input.extra_body.copy()
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/benchmarks/bench_serving.py:158]: please double check if you want to use `deepcopy(input.extra_body)` here

File: python/sglang/multimodal_gen/benchmarks/bench_serving.py [hunk 4]
```diff
@@ -718,6 +790,17 @@ async def limited_request_func(req, session, pbar):
 
     pbar.close()
 
+    # Stop profiler if it was started
+    if args.profile:
+        print("Stopping profiler and saving traces...")
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/benchmarks/bench_serving.py]: ditto

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 5]
```diff
@@ -21,6 +23,11 @@
 )
 from sglang.multimodal_gen.runtime.scheduler_client import async_scheduler_client
 from sglang.multimodal_gen.runtime.server_args import ServerArgs, get_global_server_args
+from sglang.multimodal_gen.runtime.utils.common import get_bool_env_var
+from sglang.multimodal_gen.runtime.utils.logging_utils import init_logger
+from sglang.multimodal_gen.runtime.utils.profiler import SGLDiffusionProfiler
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py]: could you please use `TYPE_CHECKING` block to wrap import-statements and reduce time to `import sglang`, as shown in issue10492

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 6]
```diff
@@ -92,6 +99,188 @@ async def health_generate():
...
+            content="Profiling is already in progress. Call /stop_profile first.\n",
+            status_code=400,
+        )
+
+    try:
+        # Get parameters
+        if obj is None:
+            obj = ProfileReqInput()
+
+        output_dir = obj.output_dir or os.getenv("SGLANG_TORCH_PROFILER_DIR", "./logs")
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py]: we already have `SGLANG_TORCH_PROFILER_DIR` in the [environ](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/environ.py#L203), please use `envs.SGLANG_TORCH_PROFILER_DIR.get()` instead of `os.getenv`

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 7]
```diff
@@ -92,6 +99,188 @@ async def health_generate():
...
+
+        # Generate unified profile_id (similar to LLM implementation)
+        profile_id = str(int(time_module.time()))
+
+        # Read env vars for with_stack and record_shapes
+        env_with_stack = get_bool_env_var("SGLANG_PROFILE_WITH_STACK", "false")
+        env_record_shapes = get_bool_env_var("SGLANG_PROFILE_RECORD_SHAPES", "false")
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py]: ditto

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 8]
```diff
@@ -92,6 +99,188 @@ async def health_generate():
...
+
+        with_stack = obj.with_stack if obj.with_stack is not None else env_with_stack
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py]: maybe we can use `with_stack = obj.with_stack or env_with_stack`

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 9]
```diff
@@ -92,6 +99,188 @@ async def health_generate():
...
+        record_shapes = (
+            obj.record_shapes if obj.record_shapes is not None else env_record_shapes
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py]: `record_shapes = obj.record_shapes or env_record_shapes`

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 10]
```diff
@@ -92,6 +99,188 @@ async def health_generate():
...
+        )
+
+        # 1. Start profiler in HTTP Server process
+        from sglang.multimodal_gen.runtime.distributed.parallel_state import (
+            get_world_rank,
+        )
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py]: from my side, this is not a good code style, could you please move this import-statement to the top of the file

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 11]
```diff
@@ -92,6 +99,188 @@ async def health_generate():
...
+
+        try:
+            rank = get_world_rank()
+        except Exception:
+            rank = 0
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py:227]: maybe we need a warning message here

File: python/sglang/multimodal_gen/runtime/entrypoints/http_server.py [hunk 12]
```diff
@@ -92,6 +99,188 @@ async def health_generate():
...
+            record_shapes=record_shapes,
+            is_host=True,
+        )
+        _global_profiler_state["profiler"] = http_profiler
+        _global_profiler_state["profile_id"] = profile_id
+
+        # 2. Start profiler in GPU Worker process via ZMQ
+        from sglang.multimodal_gen.runtime.entrypoints.openai.utils import (
+            StartProfileReq,
+        )
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/entrypoints/http_server.py]: ditto

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 13]
```diff
@@ -4,6 +4,7 @@
 import torch
 
 from sglang.multimodal_gen.runtime.utils.logging_utils import CYAN, RESET, init_logger
+from sglang.srt.utils import get_bool_env_var
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py]: pefer `Env.get()` to `get_bool_env_var`

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 14]
```diff
@@ -25,30 +25,50 @@ def __init__(
...
+        self.log_dir = log_dir or os.getenv("SGLANG_TORCH_PROFILER_DIR", "./logs")
+
+        # Read from environment variables, allow parameter override
+        env_with_stack = get_bool_env_var("SGLANG_PROFILE_WITH_STACK", "false")
+        env_record_shapes = get_bool_env_var("SGLANG_PROFILE_RECORD_SHAPES", "false")
+
+        self.with_stack = with_stack if with_stack is not None else env_with_stack
+        self.record_shapes = (
+            record_shapes if record_shapes is not None else env_record_shapes
+        )
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py]: ditto

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 15]
```diff
@@ -77,6 +97,49 @@ def __init__(
...
+    def _resolve_activities(
+        activities: list[str] | None,
+    ) -> list[torch.profiler.ProfilerActivity]:
+        """
+        Resolve user-provided activity strings to torch profiler activities.
+        """
+
+        def _default() -> list[torch.profiler.ProfilerActivity]:
+            ret = [torch.profiler.ProfilerActivity.CPU]
+            if torch.cuda.is_available():
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py]: could you please use `current_platform.is_cuda_alike` instead of `torch.cuda`

File: python/sglang/multimodal_gen/runtime/utils/profiler.py [hunk 16]
```diff
@@ -122,13 +185,16 @@ def _export_trace(self):
...
+
+            # Filename format:
+            # - Host process: {profile_id}-host.trace.json.gz
+            # - GPU Worker:   {profile_id}-rank-{rank}.trace.json.gz
+            if self.is_host:
+                filename = f"{self.request_id}-host.trace.json.gz"
+            else:
+                filename = f"{self.request_id}-rank-{self.rank}.trace.json.gz"
+
+            trace_path = os.path.abspath(os.path.join(self.log_dir, filename))
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/utils/profiler.py]: maybe we can use pathlib simplify it 
```python
from pathlib import Path

trace_path = Path(self.log_dir, filename).resolve()
```

**Comment by ping1jing2** [comment]
@Makcum888e please help to review this PR if you are free

File: python/sglang/multimodal_gen/benchmarks/bench_serving.py (hunk 3 above)
Comment (Jasen2201) [reply python/sglang/multimodal_gen/benchmarks/bench_serving.py:158]: Thank you for your review. I have double-checked and confirmed that input.extra_body.copy() suffices, as only top-level dict keys are modified.

**Review by ping1jing2 — APPROVED** [review]

File: python/sglang/multimodal_gen/benchmarks/bench_serving.py [hunk 17]
```diff
@@ -856,12 +939,30 @@ async def limited_request_func(req, session, pbar):
...
+        default=None,
+        help="Classifier-free guidance scale",
+    )
     parser.add_argument(
         "--output-file", type=str, default=None, help="Output JSON file for metrics."
     )
     parser.add_argument(
         "--disable-tqdm", action="store_true", help="Disable progress bar."
     )
+    parser.add_argument(
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/benchmarks/bench_serving.py:830]: As I can see now for profiling you need to first run manually warmup and then profiling. is it possible to do this automatically? If it needs to be done manually, please add a description so that for everyone it was clear. Maybe here?

Comment (Jasen2201) [reply python/sglang/multimodal_gen/benchmarks/bench_serving.py:830]: > As I can see now for profiling you need to first run manually warmup and then profiling. is it possible to do this automatically? If it needs to be done manually, please add a description so that for everyone it was clear. Maybe here?

Thanks for your review, I have add the **--warmup-requests** flag by referring to the implementation of bench_serving for LLM.

Comment (Makcum888e) [reply python/sglang/multimodal_gen/benchmarks/bench_serving.py:830]: Should set the default value to 1 like in `bench_serving` for LLM?

Comment (Jasen2201) [reply python/sglang/multimodal_gen/benchmarks/bench_serving.py:830]: > Should set the default value to 1 like in `bench_serving` for LLM?

Done

**Review by Makcum888e — APPROVED** [review]

**Comment by zhyajie** [comment]
I rebased on upstream/main and resolved some conflicts.

**Comment by zhyajie** [comment]
@ping1jing2   I’ve rebased to the latest main branch, fixed the conflicts, and tested the features locally. Please help me trigger the CI. Thanks a lot

**Comment by zhyajie** [comment]
@ping1jing2  I fixed the conflicts and finished the ci test.   Please take a look when you have time. Thanks

**Comment by JasonHe-WQ** [comment]
Hi, any update? I also encountered this
